### DISCUSSION TITLE
1D linear advection

### DISCUSSION THREAD

**Discussion by Toomyalla** [firedrakeproject/firedrake#2650] [Firedrake support]
Hello, 

I am trying to change the code of time-dependent  DG linear advection using unwinding [https://www.firedrakeproject.org/demos/DG_advection.py.html](url)
 to solve the problem in 1D and apply Euler for time-updating.

the initial condition is 
q(x,0)= x for x<0 .5 and 1-x otherwise.
the boundary condition is q(0,t)=0 
and the velocity, u=1

-------------------------------------------

```python3
from firedrake import *
import matplotlib.pyplot as plt

mesh= UnitIntervalMesh(100)
V = FunctionSpace(mesh, "DQ",0 )

x = SpatialCoordinate(mesh)
u =Constant(1.0)

q = Function(V).interpolate( min_value(x,1-x))
q_init = Function(V).assign(q)

qs=[]

T = 0.5
dt = T/500
dtc = Constant(dt)
q_in = Constant(0.0)


dq_trial = TrialFunction(V)
phi = TestFunction(V)
a = phi*dq_trial*dx

L = dtc*(q*div(phi*u)*dx
          - conditional(u < 0, phi*u*q_in, 0.0)*ds
          - conditional(u > 0, phi*u*q, 0.0)*ds
          - (phi('+') - phi('-'))*(un('+')*q('+') - un('-')*q('-'))*dS)

dq=Function(V)
solv=solve(a == L, dq)
t = 0.0
step = 0
output_freq = 20
while t < T - 0.5*dt:
    solv.solve()
    q.assign(q+dq)

    if step % output_freq == 0:
        qs.append(q.copy(deepcopy=True))
        print("t=", t)
```
---------------------------------------------

I got the error (DQ is supported, but handled incorrectly)
could you please let me know where the problem is? 

Thank you

**Comment by dham** [comment]
I've fixed the formatting of your code. 

Can you post the exact error please (i.e. the whole error output you see)? 

One thing that may be going on is that we usually refer to the 1D element as `DP` (i.e. we think of 1D elements as 1-simplices rather than 1-cubes) you could try to make that change and see if something changes.

**Comment by Toomyalla** [comment]
Thank you, 

UFL:ERROR Can't add expressions with different shapes.
Traceback (most recent call last):
  File "DG_linear_advection_ex1_2.py", line 11, in <module>
    q = Function(V).interpolate( min_value(x,1-x))
  File "/home/firedrake/firedrake/src/ufl/ufl/exproperators.py", line 232, in _rsub
    return Sum(o, -self)
  File "/home/firedrake/firedrake/src/ufl/ufl/algebra.py", line 41, in __new__
    error("Can't add expressions with different shapes.")
  File "/home/firedrake/firedrake/src/ufl/ufl/log.py", line 158, in error
    raise self._exception_type(self._format_raw(*message))
ufl.log.UFLException: Can't add expressions with different shapes.

-----------------------------------------------------------------------------------------------------

I've got the same error with DP

**Reply by dham** [reply]
Ah, your issue is that `x` is a 1-vector and you're using it as a scalar. The isomorphism between 1-vectors and scalars is so natural that we don't think about it, but computer algebra systems are less good at spotting it. The easiest way to get out of this is to change line 7 to:
```python
x, = SpatialCoordinate(mesh)
```
The comma after the `x` has the effect of unpacking the vector so `x` now becomes the scalar first component of the coordinates, which is what you intended.

**Comment by Toomyalla** [comment]
Thank you,  I  put the comma, and it seems to solve the issue.

Now, I get a different error :)

dq=Function(V)
    ^
SyntaxError: invalid syntax

**Reply by connorjward** [reply]
This error tends to suggest that there's in fact a syntax error on the line above where the error is thrown. I'm guessing you're missing a bracket in your complicated expression for `L`.

**Comment by Toomyalla** [comment]
Thank you all
